Fixed phishing vulnerability in HtmlHelper.php

tyspen opened the pull request:
I recently came across an article that describes a phishing vulnerability. Instagram has already fixed the vulnerability on their site and I feel Cake users should do the same on their projects. Consider adding a configuration option to disable this feature due to using it to find the parent window.

Most users won't need this feature to be disabled, and only should be disabled if the user directly knows the reasoning for doing so.

[Article here](https://dev.to/ben/the-targetblank-vulnerability-by-example)

tyspen committed:
Fixed phishing vulnerability in HtmlHelper.php

tyspen committed:
Updated lines

Added line spacing between end bracket

tyspen committed:
Stickler-ci fix

Spacing fix

jadb commented on src/View/Helper/HtmlHelper.php:
```diff
@@ -344,7 +344,14 @@ public function link($title, $url = null, array $options = [])
+        if (isset($options['target'])) {
+            $options['rel'] = 'noopener noreferrer';
+        }
+        
```
This will override user's defined `rel` in certain cases (i.e. bookmark, author, etc.).

josegonzalez replied:
We also don't add comments inline in code - it's a smell and usually means we need to be more clear about the intentions of the code.

tspencer9303 replied:
So removing inline comments and adding user's defined `rel` with it?

jadb replied:
I think so, yes :)

tspencer9303 replied:
My thoughts of fixing this would be adding an option in the link such as `"targetProtect" => false` With this, I feel it best if I make a private function to handle this, inside of the HtmlHelper. Would that be appropriate or should I go about it in a different manner?

markstory replied:
If `rel` is defined we shouldn't be replacing it automatically under any circumstance. But I echo @jadb and @josegonzalez 's comments.

jadb replied:
Maybe more something like:

```php
$options += ['rel' => ''];
$current = explode(' ', $options['rel']);
$nophishing = ['noopener', 'noreferrer'];
$options['rel'] = implode(' ', array_filter(array_unique(array_merge($current, $nophishing))));
```

tspencer9303 replied:
If `rel` is set, then the user is specifically assigning a relationship between the current page, and the one it's linked to. So after seeing your comment, it would be impractical for the user to define `rel` and also disable the protection.

jadb replied:
The code I shared above was copied from a custom helper where I had this automation implemented ;)

It will only append to the user's defined `rel`.

jadb commented:
While I like the idea, it needs some more work and tests to at least cover the case I highlighted. 

Not sure what others think

josegonzalez commented:
👍 on the idea, with a few changes and a test, I think it should be merged :)

tyspen committed:
Updated HtmlHelper

Removed comment lines that were not needed and changed the `$options['rel']` to only apply `noopener noreferrer` if `rel` has not been defined by the user. The reasoning for this is because it would be impractical for the user to explicitly disable the feature if it is implied by assigning a relationship in the first place.

jadb commented:
Thanks for updating it. Still needs testing though.

Also adding this for posterity since it was collapsed (slightly improved):

```php
$options += ['rel' => ''];
$current = is_array($options['rel']) ? $options['rel'] : explode(' ', $options['rel']);
$nophishing = ['noopener', 'noreferrer'];
$options['rel'] = implode(' ', array_filter(array_unique(array_merge($current, $nophishing))));
```

jadb commented:
Should this just be for `target='_blank'`?

chinpei215 commented:
Interesting. I think it would be more better should check whether `_target` attribute really contains an external domain or not.
@jadb Probably no.

jadb commented:
@chinpei215, I'd like to add to my above reactions, wouldn't this become a little more complicated though? 

And mind (given your security expertise) elaborate a little more on why the other values might be at risk (not just `_blank`)? 

I ask because then, since `_self` is the default, all links on the web should contain `noopener noreferrer`, no?

ADmad commented:
@jadb @chinpei215 https://medium.com/@Jamie_Farrelly/browsers-are-broken-but-nobody-cares-all-it-took-was-1-line-of-code-to-fix-it-f8af13c18cff#.g0p28wpur

So it seems the vulnerability only exists for `target="_blank"`

chinpei215 commented:
@ADmad I'll try to check in my local environment.
And sorry. I wanted say "check whether **href** attribute really contains an external domain or not" in the above comment.

ADmad commented on src/View/Helper/HtmlHelper.php:
```diff
@@ -344,7 +344,11 @@ public function link($title, $url = null, array $options = [])
             $escapeTitle = false;
         }
-
+        
```
Extra whitespace

ADmad replied:
@antograssiot Could you please check if we can add a sniff for trailing whitespace to our ruleset?

antograssiot replied:
I'll try to have a look next we

jadb commented:
Link at the bottom of article linked by @ADmad seems like the best source IMO:

https://mathiasbynens.github.io/rel-noopener/

Can't go wrong with Mathias!

All that been said, am not so sure anymore this should be covered at a framework level. The JS fix for it is trivial and without JS-enabled browsers, there is no vulnerability (to my understanding).

chinpei215 commented:
@jadb @ADmad I checked in my local environment. It seems the vulnerability not only exists for target="_blank", but also target="something" (a named window).

jadb commented:
So to resume, it would need to add `noopener` (and `noreferrer` for older browsers) only if:

- link is outgoing
- `target` is `_blank` or not `_self`, `_parent`, `_top`

Right @chinpei215?

chinpei215 commented:
@jadb Probably right. But I am not familiar with this attack. So I might be wrong.

chinpei215 commented:
Ah but we can wheel-click all links...

ravage84 commented:
Actually, IMHO this should be handled by the main web browser vendors, not by PHP server-side frameworks.

Just like the rise of web browser built-in pop up blocking.

tspencer9303 commented:
Honestly it should be, but currently it's not and the issue has been around for years. It's enough for larger sites to take action against, so I believe it's something Cake should take action against as well. Not all users are aware of it and could be effected by it if they allow user provided links.

dereuromark commented:
I think if this is provided as an opt-in feature this would be fine. It just shouldnt be the default, not only for BC reasons.

ADmad commented:
Yes this feature should be implemented as a HtmlHelper setting and should be opt-in.

dereuromark commented:
Is anyone willing to provide the modified version as PR so we can implement this security feature?

markstory commented:
Closing as in the current state I'm not comfortable changing the default behavior of HtmlHelper in a possibly backwards incompatible way.
